Phpstan parser

patrickkusebauch opened the pull request:
Replaces https://github.com/deptrac/deptrac-src/pull/116

patrickkusebauch committed:
Revival of PHPStan parser

patrickkusebauch committed:
Resolve bundled dependency mismatch.

Code style.

patrickkusebauch committed:
Proper assignment of code to modules.

patrickkusebauch committed:
Consistent Docs parsing and resolving some CI issues.

gennadigennadigennadi commented:
Is this one ready for review? Do you plan to change something? If not I'll try to find a moment next week and go through the changes.

Changing the minimal php version to 8.2 is necessary, I haven't checked yet? In the past we only dropped support for eol version. 8.1 still gets security fixes till the end of 2025.

patrickkusebauch commented:
Yes, it is ready for review. The change of minimal version is necessary. It is because of upgrading `phpstan/phpdoc-parser` to v2 and the related compatibility issues. And I needed to do this upgrade to use the same version as is bundled with `phpstan/phpstan` so that the connectors to it work properly. Much better version of this upgrade (IMHO) is in https://github.com/deptrac/deptrac/pull/1485. I did just the bare minimum required.

gennadigennadigennadi commented:
Could you point me to the compatibility issues or do you have a reproducer at hand? I see `phpstan/phpdoc-parser` and `phpstan/phpstan` **not** requiring php8.2. phpstan-src does requirer php 8.2 but as I understand it gets downgraded to 7.4|8.0 for the bundle version.

Does the issue occur if you use  phpstan/phpdoc-parser and phpstan/phpstan and if yes what are the conflicting versions?

gennadigennadigennadi commented:
okay something if seen so far is, that `phpstan/phpstan-src` (2.1.x) requires an explizit phpdoc-parser version (`2.2`), but `phpstan/phpstan` does not. So we should be fine with php 8.1?

And if it`s important for this PR to work correctly, that our `phpdoc-parser` version and `phpstan/phpstan` scoped `phpd-parser` are in sync, its just a question of time for deptrac to break (at least this experimental feature). Because we would have to keep track of phpstan dependencies?

gennadigennadigennadi commented:
@patrickkusebauch: I just changed the php version from composer.json to `8.1` and everything seems to work well (based on you phpstan-parser branch). The min. php version is not really a blocker, but I'm confused because I don't see were we would need to increase the min. php version yet. 

<img width="507" height="566" alt="Screenshot 2025-08-12 at 20 25 17" src="https://github.com/user-attachments/assets/e9c9762f-17e0-4605-8e12-3523adb20eeb" />

But also this branch needs to resolve its conflicts.

patrickkusebauch commented:
I am busy for the rest of the week. Will look into it next week.

patrickkusebauch committed:
Rebase update

patrickkusebauch committed:
Fix GitHub Action caching and force tools to use the minimal supported version for installation.

patrickkusebauch committed:
Add PHP 8.4 coverage to some GitHub Actions.

Force infection to use at least 8.2 because of incompatibility between the supported PHP versions of the infection plugin and nikic PHP parser.

```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - infection/infection 0.27.10 requires nikic/php-parser ^4.15.1 -> found nikic/php-parser[v4.15.1, ..., v4.19.4] but it conflicts with your root composer.json require (^5).
    - roave/infection-static-analysis-plugin[1.36.0, ..., 1.39.0] require php ~8.2.0 || ~8.3.0 || ~8.4.0 -> your php version (8.1; overridden via config.platform, actual: 8.3.6) does not satisfy that requirement.
    - roave/infection-static-analysis-plugin 1.35.0 requires infection/infection 0.27.10 -> satisfiable by infection/infection[0.27.10].
    - Root composer.json requires roave/infection-static-analysis-plugin ^1.35 -> satisfiable by roave/infection-static-analysis-plugin[1.35.0, ..., 1.39.0].
```

patrickkusebauch reviewed (commented):
Ready for review. @gennadigennadigennadi

gennadigennadigennadi commented on src/DefaultBehavior/Ast/DocParsingHelper.php:
```diff
@@ -0,0 +1,65 @@
+
+        $fileTypeMapper = $phpStanContainer->createFileTypeMapper();
+        $classReflection = $scope->getClassReflection();
+        assert(null !== $classReflection);
```
This crashes in the following case.

<img width="564" height="447" alt="Image" src="https://github.com/user-attachments/assets/0866d0e6-8688-46aa-bcf3-e115b73d2f0b" />

It seems that if the scope is a trait `getClassReflection` does not return something.

gennadigennadigennadi replied:
https://github.com/phpstan/phpstan-src/blob/d78a5d86c3c00a62790be81d65355d4ca344ff39/src/Analyser/MutatingScope.php#L398

Checking isInClass before probably would fix the problem. And for Traits isInTrait.

patrickkusebauch replied:
Resolved.

patrickkusebauch committed:
Fix trait docs parsing.

gennadigennadigennadi commented:
@patrickkusebauch there seems to be a merge conflict? Could you resolve it and I would merge this PR afterwards.

gennadigennadigennadi commented on tests/Core/Ast/Parser/NikicTypeResolverTest.php:
This should also be renamed?

patrickkusebauch committed:
Test rename.

patrickkusebauch commented:
@gennadigennadigennadi done. Rebased and merge conflict resolved.
